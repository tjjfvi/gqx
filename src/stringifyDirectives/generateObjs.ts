
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
