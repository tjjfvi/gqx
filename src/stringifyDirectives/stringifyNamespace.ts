import { Obj } from ".";
import stringifyValue from "./stringifyValue";
import indent from "../indent";

export default (objs: Obj[]) =>
  `export namespace $$Directives {\n${
    objs.map(obj =>
      indent`export namespace ${obj.type} {\n${
        obj.props.map(p =>
          indent`export type ${p.prop} = {${
            p.dirs.length ?
              `\n${
                p.dirs.map(dir =>
                  indent`${dir.dirName}: {${
                    dir.dirArgs && dir.dirArgs.length ?
                      indent`\n${
                        dir.dirArgs.map(arg =>
                          `${arg.name.value}: ${stringifyValue(arg.value)},`
                        ).join("\n")
                      }\n` :
                      ""
                  }},`
                ).join("\n")
              }\n` :
              ""
          }}`
        ).join("\n")
      }\n}`
    ).join("\n")
  }\n${
    indent`export type $<T> =${objs.map(o => o.props.map(p =>
      indent`\nT extends ${p.type}$${p.prop} ? ${p.type}.${p.prop} :`
    ).join("")).join("")}\n${indent`never\n`}`
  }}`
