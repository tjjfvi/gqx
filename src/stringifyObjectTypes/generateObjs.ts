
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
