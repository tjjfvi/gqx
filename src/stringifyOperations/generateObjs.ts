
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
