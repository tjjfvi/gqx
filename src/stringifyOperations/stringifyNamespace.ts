import { Obj } from "."
import stringifyId from "./stringifyId"
import { Prop } from ".";

export default (
  name: string,
  objs: Obj[],
  f: (obj: Obj, prop: Prop) => string,
  extraGenerics = "",
  cond = (o: Obj, p: Prop) =>
    `    T extends typeof ${stringifyId(p.id)} ? ${o.type}.${p.id.prop} :`
) =>
  `
export namespace ${name} {
${
  objs.map(o =>
    `  export namespace ${o.type} {\n${o.props.map(p =>
      `    export ` + f(o, p) + `\n`
    ).join("")}  }\n`
  ).join("")
}
  export type $<T${extraGenerics}> =
${
  objs.map(o => o.props.map(p =>
    cond(o, p)
  ).join("\n")).join("\n")
}
    never
}
`.trim()
