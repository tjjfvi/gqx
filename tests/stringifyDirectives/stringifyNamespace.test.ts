import stringifyNamespace from "../../src/stringifyDirectives/stringifyNamespace"
import generateObjs from "../../src/stringifyDirectives/generateObjs"

test("", () => {
  let code = stringifyNamespace(generateObjs(utils.getOperations(["stringifyDirectives"])));
  expect(code).toParse().toMatchSnapshot();
})
