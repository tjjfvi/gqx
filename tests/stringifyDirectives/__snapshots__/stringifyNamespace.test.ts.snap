// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"export namespace $$Directives {
  export namespace Query {
    export type noDirectives = {}
    export type oneDirective = {
      scalarArg: {
        x: \\"string\\",
      },
    }
    export type multDirectivesArgs = {
      scalarArg: {
        a: \\"test\\",
        b: 12345,
      },
      inputArg: {
        a: {
          x: 5,
          y: [
            1,
            2,
            3,
          ],
          z: {
            x: 5,
          },
        },
      },
      scalarArrArg: {
        a: [
          \\"a\\",
          \\"b\\",
          \\"c\\",
          null,
        ],
      },
      inputArrArg: {
        a: [
          {
            x: 5,
          },
          {
            x: 6,
          },
          {
            y: [
              1,
              2,
              3,
            ],
          },
        ],
      },
    }
    export type multDirectivesNullArgs = {
      scalarArg: {
        a: null,
        b: null,
      },
      inputArg: {
        a: null,
      },
      scalarArrArg: {
        a: null,
      },
      inputArrArg: {
        a: null,
      },
    }
    export type multDirectivesNoArgs = {
      noArgs: {},
      scalarArg: {},
      inputArg: {},
      scalarArrArg: {},
      inputArrArg: {},
    }
  }
  export type $<T> =
    T extends typeof Query$noDirectives ? Query.noDirectives :
    T extends typeof Query$oneDirective ? Query.oneDirective :
    T extends typeof Query$multDirectivesArgs ? Query.multDirectivesArgs :
    T extends typeof Query$multDirectivesNullArgs ? Query.multDirectivesNullArgs :
    T extends typeof Query$multDirectivesNoArgs ? Query.multDirectivesNoArgs :
    never
}"
`;
